src/commands/parser.rs

```diff
@@ -1,9 +1,196 @@
-use nom::{*, bytes::complete::{take_until, take_till}, character::{is_newline}, error::*, sequence::*};
+use nom::{
+    bytes::complete::{take_until, tag},
+    character::complete::{newline, not_line_ending},
+    combinator::opt,
+    multi::many0,
+    *,
+};
 
-pub fn todo() -> impl Fn(&[u8]) -> IResult<&[u8], &[u8], Error<&[u8]>> {
-    move |i: &[u8]| {
-	let header = take_until("TODO:");
-	let content = take_till(is_newline);
-	preceded(header, content)(i)
+#[derive(Debug, PartialEq, Clone)]
+pub struct Todo {
+    pub tag: String,
+    pub text: String,
+}
+
+fn parse_todo<'a>(todo_tag: &'a str, input: &'a str) -> IResult<&'a str, Todo> {
+    // remove everything prior to the tag
+    let (input, _) = take_until(todo_tag)(input)?;
+
+    let (input, _) = tag(todo_tag)(input)?;
+    let (input, text) = not_line_ending(input)?;
+
+    // discard optional line ending
+    let (input, _) = opt(newline)(input)?;
+
+    Ok((input, Todo {
+        tag: todo_tag.to_owned(),
+        text: text.to_owned(),
+    }))
+}
+
+fn parse_todos(file_content: &str) -> IResult<&str, Vec<Todo>> {
+    let line_parser = | file_content | parse_todo("TODO:", file_content);
+    let (input, todos) = many0(line_parser)(file_content)?;
+    Ok((input, todos))
+}
+
+pub fn parse_file(file_content: &str) -> Vec<Todo> {
+    match parse_todos(file_content) {
+        Ok((_, todos)) => todos,
+        Err(_) => vec![],
+    }
+}
+
+
+#[cfg(test)]
+mod test {
```

Comment:
I think tests should not be in the same files as normal code. There's a chance this is a Rust convention though. If so, ignore this comment (I will be pretty sad if this is the case).

Reply:
Rust convention: https://doc.rust-lang.org/book/ch11-03-test-organization.html#unit-tests